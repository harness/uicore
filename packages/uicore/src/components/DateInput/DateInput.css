/*
 * Copyright (c) Harness Inc.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

.dateInput {
  position: relative;
}

.calendarIcon {
  position: absolute;
  top: 1px;
  right: 1px;
  button {
    height: 30px !important;
    width: 30px !important;
  }
}

.helpText {
  font-size: var(--font-size-small);
  margin-top: var(--spacing-small);
}
.helpIcon {
  margin-left: var(--spacing-small);
}
