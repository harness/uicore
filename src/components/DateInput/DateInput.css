/*
 * Copyright 2020 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.dateInput {
  position: relative;
}

.calendarIcon {
  position: absolute;
  top: 1px;
  right: 1px;
  button {
    height: 30px !important;
    width: 30px !important;
  }
}

.helpText {
  font-size: var(--font-size-small);
  margin-top: var(--spacing-small);
}
.helpIcon {
  margin-left: var(--spacing-small);
}
