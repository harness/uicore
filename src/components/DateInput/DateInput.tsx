import React, { useState, useEffect } from 'react'
import css from './DateInput.css'
import { TextInput, TextInputProps } from '../TextInput/TextInput'
import { DatePicker, TimePrecision, IDatePickerProps } from '@blueprintjs/datetime'
import { Intent } from '@blueprintjs/core'
import { Button } from '../Button/Button'
import { Popover, PopoverProps } from '../Popover/Popover'
import {
  getHelpPopoverContent,
  DurationInputHelpers,
  parseStringToTime,
  timeToDisplayText
} from '../DurationInput/DurationInput'
import moment from 'moment'
import { Icon } from '../../icons/Icon'

export interface DateInputProps extends Omit<TextInputProps, 'onChange'> {
  dateTimeFormat?: string
  dateProps?: IDatePickerProps
  timePrecision?: TimePrecision
  onChange?: (value: string | undefined, error?: string) => void
  allowVariables?: boolean
  popoverProps?: PopoverProps
}
enum Errors {
  NOT_VALID_DATE = 'Not a valid date',
  NOT_VALID_EXPRESSION = 'Not a valid Date Expression'
}
interface DateInputState {
  isDate: boolean
  value?: string
  intent?: Intent
  error?: Errors | undefined
}

const { UNITS_ORDER: ALL_UNITS } = DurationInputHelpers

const REGEX_VALID_TEXT = /^(current\(\)|\${[a-z\d._]+})(\s*[+-]\s*)*([0-9]+[wdhms]+\s*)*$/i
const REGEX_VALID_MS_TEXT = /^(current\(\)|\${[a-z\d._]+})(\s*[+-]\s*)*([0-9]+)*$/i
const REGEX_VALID_VARIABLE = /^(current\(\)|\${[a-z\d._]+})(\s*[+-]\s*)*/i
const REGEX_VALID_VALUE = /([0-9]+[wdhms]+\s*)*$/i
const REGEX_VALID_MS_VALUE = /([0-9])*$/i
const REGEX_VALID_DATE = /^\d{2,4}.\d{2,4}.\d{2,4}$/i
const REGEX_VALID_DATE_TIME = /^\d{2,4}.\d{2,4}.\d{2,4}\s\d{1,2}:\d{2}$/i

const isValidDateInput = (
  value: number | string | undefined,
  minDate: Date,
  maxDate: Date,
  formatDateTime: string,
  isTime: boolean
) => {
  if (value === '') {
    return true
  }
  const isRegExValid =
    typeof value === 'number' ||
    (typeof value === 'string' && (isTime ? REGEX_VALID_DATE_TIME.test(value) : REGEX_VALID_DATE.test(value)))
  const date = moment(value, formatDateTime)
  return isRegExValid && date.isValid() && date.isBetween(minDate, maxDate)
}

export const DateInput: React.FC<DateInputProps> = props => {
  const { timePrecision, allowVariables, dateProps, popoverProps, ...otherProps } = props
  const { dateTimeFormat, onChange } = props
  const formatDateTime = dateTimeFormat || (timePrecision ? 'MM/DD/YYYY HH:mm' : 'MM/DD/YYYY')

  const maxDate = dateProps && dateProps.maxDate ? dateProps.maxDate : new Date(2050, 5, 5)
  const minDate = dateProps && dateProps.minDate ? dateProps.minDate : new Date(1990, 5, 5)

  const [state, setState] = useState<DateInputState>({ isDate: true, value: '' })
  let dateDefaultValue = new Date()

  useEffect(() => {
    const { value, allowVariables } = props
    let displayValue = ''
    if (value && typeof value === 'string' && value !== '')
      if (!isNaN(value as any)) {
        // Check long number
        displayValue = moment(parseInt(value)).format(formatDateTime)
        dateDefaultValue = new Date(parseInt(value))
        setState(prevState => ({
          ...prevState,
          value: displayValue,
          error: undefined,
          isDate: true
        }))
      } else if (allowVariables) {
        // If allowVariables set to True
        const isValid = REGEX_VALID_MS_TEXT.test(value)
        if (isValid) {
          // Valid value
          const timeValueString = value.replace(REGEX_VALID_VARIABLE, '').trim()
          const timeDisplay =
            timeValueString && timeValueString !== '' && !isNaN(timeValueString as any)
              ? timeToDisplayText(parseInt(timeValueString))
              : ''
          displayValue = value.replace(REGEX_VALID_MS_VALUE, '') + timeDisplay
          setState(prevState => ({
            ...prevState,
            value: displayValue,
            isDate: false,
            error: undefined,
            intent: Intent.SUCCESS
          }))
        } else {
          // Invalid Value
          setState(prevState => ({
            ...prevState,
            value,
            error: Errors.NOT_VALID_EXPRESSION,
            isDate: false,
            intent: Intent.DANGER
          }))
        }
      } else {
        // Fallback to see if moment can convert this to date/datetime
        const isValidDate = isValidDateInput(value, minDate, maxDate, formatDateTime, !!timePrecision)
        setState(prevState => ({
          ...prevState,
          value,
          isDate: isValidDate,
          error: Errors.NOT_VALID_DATE,
          intent: isValidDate ? undefined : Intent.DANGER
        }))
      }
  }, [props.value])

  // OnChange of Input Box
  const onChangeText = (event: React.ChangeEvent<HTMLInputElement>) => {
    const originalValue = event.target.value
    const value = event.target.value.trim()
    const isValidDate = isValidDateInput(value, minDate, maxDate, formatDateTime, !!timePrecision)
    const isValidText = REGEX_VALID_TEXT.test(value)
    const isCustomValue = REGEX_VALID_VARIABLE.test(value)

    // Check if OnChange is passes as props
    if (onChange) {
      // Valid Input
      if (isValidText && isCustomValue) {
        // Extract time from string
        const time = parseStringToTime(value.replace(REGEX_VALID_VARIABLE, '').trim())
        const timeString = time !== 0 ? time.toString() : ''
        onChange(value.replace(REGEX_VALID_VALUE, '') + timeString)
      } else if (isValidDate && !isCustomValue) {
        if (value !== '') {
          onChange(moment(value, formatDateTime).valueOf().toString())
        } else {
          onChange(value)
        }
      } else if (isCustomValue) {
        // Not a Valid Date expression
        onChange(undefined, Errors.NOT_VALID_EXPRESSION)
      } else {
        // Not a Valid Date
        onChange(undefined, Errors.NOT_VALID_DATE)
      }
    }

    // Update State
    if (isValidText) {
      setState(prevState => ({
        ...prevState,
        value: originalValue,
        isDate: false,
        intent: Intent.SUCCESS,
        error: undefined
      }))
    } else if (isValidDate && !isCustomValue) {
      setState(prevState => ({
        ...prevState,
        value: originalValue,
        isDate: isValidDate,
        intent: undefined,
        error: undefined
      }))
    } else {
      if (isCustomValue) {
        setState(prevState => ({
          ...prevState,
          value: originalValue,
          isDate: false,
          error: Errors.NOT_VALID_EXPRESSION,
          intent: Intent.DANGER
        }))
      } else {
        setState(prevState => ({
          ...prevState,
          value: originalValue,
          isDate: isValidDate,
          error: Errors.NOT_VALID_DATE,
          intent: Intent.DANGER
        }))
      }
    }
  }

  // On Change with CAlender
  const onChangeDate = (selectedDate: Date) => {
    if (onChange) {
      onChange(selectedDate.getTime().toString())
    }
    setState(prevState => ({ ...prevState, value: moment(selectedDate.getTime()).format(formatDateTime) }))
  }

  // Pase Date input pass to blueprint DatePicker
  const dateValue = state.value && state.isDate ? moment(state.value, formatDateTime).toDate() : undefined
  const dateInputProps = {
    defaultValue: dateDefaultValue,
    value: dateValue,
    closeOnSelection: false,
    popoverProps: {
      usePortal: false
    },
    formatDate: (date: Date) => moment(date.getTime()).format(formatDateTime),
    parseDate: (str: string) => new Date(str),
    maxDate,
    minDate,
    placeholder: formatDateTime,
    timePrecision,
    onChange: onChangeDate,
    ...dateProps
  }

  return (
    <div className={css.dateInput}>
      <TextInput
        {...otherProps}
        type="text"
        value={state.value}
        intent={state.intent}
        placeholder={formatDateTime}
        errorText={state.error}
        onChange={onChangeText}
      />
      {state.isDate ? (
        <Popover
          boundary="viewport"
          usePortal={false}
          position="top-left"
          className={css.calendarIcon}
          content={<DatePicker {...dateInputProps} />}
          {...popoverProps}>
          <Button minimal icon="calendar" />
        </Popover>
      ) : null}
      {allowVariables ? (
        <div className={css.helpText}>
          Support custom variables: &#x22;current() + 2d 2h&#x22;
          <Popover
            wrapperTagName="span"
            className={css.helpIcon}
            content={getHelpPopoverContent(ALL_UNITS)}
            lazy={true}
            interactionKind="hover"
            position="top"
            usePortal={false}
            {...popoverProps}>
            <Icon name="question" size={14} />
          </Popover>{' '}
        </div>
      ) : null}
    </div>
  )
}
