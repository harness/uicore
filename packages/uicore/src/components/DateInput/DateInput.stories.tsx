/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/* eslint-disable no-console */
import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary, Description } from '@storybook/addon-docs/blocks'
import { DateInput } from '../..'
import { DateInputProps } from '../DateInput/DateInput'
export default {
  title: 'Form / DateInput',

  component: DateInput,
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>DateInput</Title>
            <Subtitle>
              <pre>
                <code>{`import {DateInput} from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>
            <Description>{`DateInput use cases

- Show date and date time and return response in milliseconds in strings
- Use Calender for date and date time selection
- Uses same props as TextInput
- Support for custom variables like current() & \${variable} workflow variables
  - Plus/Minus Weeks/Days/Hours/Minutes/Seconds similar to DurationInput
  - Returns the response as current() \${variable} +- duration
- Show valid date input`}</Description>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta
export const ShowDate: Story<DateInputProps> = args => {
  return <DateInput {...args} />
}
export const ShowTime: Story<DateInputProps> = args => {
  return <DateInput timePrecision="minute" {...args} />
}
export const ShowDateWithValue: Story<DateInputProps> = args => {
  return <DateInput value="1579773229467" {...args} />
}
export const ShowTimeWithValue: Story<DateInputProps> = args => {
  return <DateInput timePrecision="minute" value="1579773229467" {...args} />
}
export const ShowDatewithCustomVariables: Story<DateInputProps> = args => {
  // eslint-disable-next-line no-console
  return <DateInput allowVariables={true} onChange={(value, error) => console.log(value, error)} {...args} />
}
export const ShowTimeWithCustomVariables: Story<DateInputProps> = args => {
  return (
    <DateInput
      timePrecision="minute"
      allowVariables={true}
      onChange={(value, error) => console.log(value, error)}
      {...args}
    />
  )
}
export const ShowDateWithCustomVariablesWithValue: Story<DateInputProps> = args => {
  return (
    <DateInput
      allowVariables={true}
      value="current() + 7200000"
      onChange={(value, error) => console.log(value, error)}
      {...args}
    />
  )
}
export const ShowTimeWithCustomVariablesWithValue: Story<DateInputProps> = args => {
  return (
    <DateInput
      timePrecision="minute"
      allowVariables={true}
      value="${App.TIME_LOCAL} - 777600000"
      onChange={(value, error) => console.log(value, error)}
      {...args}
    />
  )
}
export const ShowDateWithCustomVariablesWithErrorValue: Story<DateInputProps> = args => {
  return <DateInput allowVariables={true} value="current() + 2h 10m + 2" {...args} />
}
export const ShowTimeWithCustomVariablesWithErrorValue: Story<DateInputProps> = args => {
  return <DateInput timePrecision="minute" allowVariables={true} value="current() + w X" {...args} />
}
