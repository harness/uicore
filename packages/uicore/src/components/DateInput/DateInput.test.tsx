/*
 * Copyright 2020 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { DateInput } from './DateInput'

describe('<DateInput/> tests', () => {
  test('should always be UTC', () => {
    expect(new Date().getTimezoneOffset()).toBe(0)
  })
  test('renders with empty value date', () => {
    const onChange = jest.fn()
    const { container, getByTestId } = render(<DateInput onChange={onChange} data-testid="input" />)
    expect(container).toMatchSnapshot()
    fireEvent.change(getByTestId('input'), {
      target: {
        value: '01/23/2020'
      }
    })
    expect(onChange).toHaveBeenCalledWith('1579737600000')
    // Check not a valid Date
    fireEvent.change(getByTestId('input'), {
      target: {
        value: 'asdasd'
      }
    })
    expect(onChange).toHaveBeenCalledWith(undefined, 'Not a valid date')
  })

  test('renders with empty value date time', () => {
    const onChange = jest.fn()
    const { container, getByTestId } = render(
      <DateInput timePrecision="minute" onChange={onChange} data-testid="input" />
    )
    expect(container).toMatchSnapshot()
    fireEvent.change(getByTestId('input'), {
      target: {
        value: '01/23/2020 15:23'
      }
    })
    expect(onChange).toHaveBeenCalledWith('1579792980000')
  })

  test('renders with empty value and allow variables', () => {
    const onChange = jest.fn()
    const { container, getByTestId } = render(
      <DateInput allowVariables={true} onChange={onChange} data-testid="input" />
    )
    expect(container).toMatchSnapshot()
    // Check plus
    fireEvent.change(getByTestId('input'), {
      target: {
        value: 'current() + 2h'
      }
    })
    expect(onChange).toHaveBeenCalledWith('current() + 7200000')
    // Check minus
    fireEvent.change(getByTestId('input'), {
      target: {
        value: 'current() - 2h'
      }
    })
    expect(onChange).toHaveBeenCalledWith('current() - 7200000')
  })

  test('renders with value and allow variables', () => {
    const { container } = render(<DateInput allowVariables={true} value="current() + 7200000" />)
    expect(container).toMatchSnapshot()
  })

  test('renders with invalid value and allow variables', () => {
    const { container } = render(<DateInput allowVariables={true} value="current() + 2x" />)
    expect(container).toMatchSnapshot()
  })

  test('renders with value ', () => {
    const { container } = render(<DateInput value="1579773180000" />)
    expect(container).toMatchSnapshot()
  })

  test('renders with placeholder ', () => {
    const { getByPlaceholderText } = render(
      <DateInput value="1579773180000" placeholder={'This is dateInput component'} />
    )
    expect(getByPlaceholderText('This is dateInput component')).toBeInTheDocument()
  })
})
