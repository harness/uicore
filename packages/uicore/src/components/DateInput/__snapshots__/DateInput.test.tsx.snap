// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput/> tests renders with empty value and allow variables 1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          data-testid="input"
          placeholder="MM/DD/YYYY"
          type="text"
          value=""
        />
      </div>
    </div>
    <span
      class="bp3-popover-wrapper calendarIcon"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-calendar StyledProps--main"
            icon="calendar"
          >
            <svg
              data-icon="calendar"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                calendar
              </desc>
              <path
                d="M11 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1s-1 .4-1 1v1c0 .5.4 1 1 1zm3-2h-1v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H6v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H1c-.6 0-1 .5-1 1v12c0 .6.4 1 1 1h13c.6 0 1-.4 1-1V2c0-.6-.5-1-1-1zM5 13H2v-3h3v3zm0-4H2V6h3v3zm4 4H6v-3h3v3zm0-4H6V6h3v3zm4 4h-3v-3h3v3zm0-4h-3V6h3v3zM4 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1S3 .4 3 1v1c0 .5.4 1 1 1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
    </span>
    <div
      class="helpText"
    >
      Support custom variables: "current() + 2d 2h"
      <span
        class="bp3-popover-wrapper helpIcon"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="question"
            tabindex="0"
          >
            <svg
              height="14"
              viewBox="0 0 14 14"
              width="14"
            >
              <path
                d="M7 .1C3.2.1.1 3.2.1 7s3.1 6.9 6.9 6.9 6.9-3.1 6.9-6.9S10.8.1 7 .1zm0 1.2c3.2 0 5.7 2.6 5.7 5.7s-2.6 5.7-5.7 5.7S1.3 10.2 1.3 7 3.8 1.3 7 1.3zm0 2.3c-1.3 0-2.3 1-2.3 2.3h1.1c0-.6.5-1.1 1.1-1.1S8 5.3 8 5.9c0 .4-.3.8-.7 1h-.1c-.4.2-.8.6-.8 1.1v.7h1.1V8l.2-.1c.9-.3 1.5-1.1 1.5-2.1.1-1.2-.9-2.2-2.2-2.2zm-.6 5.7v1.1h1.1V9.3H6.4z"
                fill="#9AA5B5"
              />
            </svg>
          </span>
        </span>
      </span>
       
    </div>
  </div>
</div>
`;

exports[`<DateInput/> tests renders with empty value date 1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          data-testid="input"
          placeholder="MM/DD/YYYY"
          type="text"
          value=""
        />
      </div>
    </div>
    <span
      class="bp3-popover-wrapper calendarIcon"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-calendar StyledProps--main"
            icon="calendar"
          >
            <svg
              data-icon="calendar"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                calendar
              </desc>
              <path
                d="M11 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1s-1 .4-1 1v1c0 .5.4 1 1 1zm3-2h-1v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H6v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H1c-.6 0-1 .5-1 1v12c0 .6.4 1 1 1h13c.6 0 1-.4 1-1V2c0-.6-.5-1-1-1zM5 13H2v-3h3v3zm0-4H2V6h3v3zm4 4H6v-3h3v3zm0-4H6V6h3v3zm4 4h-3v-3h3v3zm0-4h-3V6h3v3zM4 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1S3 .4 3 1v1c0 .5.4 1 1 1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
    </span>
  </div>
</div>
`;

exports[`<DateInput/> tests renders with empty value date time 1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          data-testid="input"
          placeholder="MM/DD/YYYY HH:mm"
          type="text"
          value=""
        />
      </div>
    </div>
    <span
      class="bp3-popover-wrapper calendarIcon"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-calendar StyledProps--main"
            icon="calendar"
          >
            <svg
              data-icon="calendar"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                calendar
              </desc>
              <path
                d="M11 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1s-1 .4-1 1v1c0 .5.4 1 1 1zm3-2h-1v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H6v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H1c-.6 0-1 .5-1 1v12c0 .6.4 1 1 1h13c.6 0 1-.4 1-1V2c0-.6-.5-1-1-1zM5 13H2v-3h3v3zm0-4H2V6h3v3zm4 4H6v-3h3v3zm0-4H6V6h3v3zm4 4h-3v-3h3v3zm0-4h-3V6h3v3zM4 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1S3 .4 3 1v1c0 .5.4 1 1 1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
    </span>
  </div>
</div>
`;

exports[`<DateInput/> tests renders with invalid value and allow variables 1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main hasError"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="MM/DD/YYYY"
          type="text"
          value="current() + 2x"
        />
      </div>
      <div
        class="errorDiv"
        data-name=""
      >
        <span
          class="bp3-icon StyledProps--main errorTextIcon StyledProps--intent-danger"
          data-icon="circle-cross"
        >
          <svg
            height="12"
            viewBox="0 0 13 13"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M.167 6.501a6.333 6.333 0 1012.666 0 6.333 6.333 0 00-12.666 0zm4.071-3.003l2.26 2.263 2.264-2.263c.495-.495 1.236.248.741.743L7.24 6.504l2.254 2.258c.495.495-.247 1.238-.742.743L6.497 7.246 4.238 9.505c-.495.495-1.236-.249-.741-.744l2.258-2.258-2.259-2.262c-.494-.495.248-1.238.742-.743z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="error"
        >
          Not a valid Date Expression
        </span>
      </div>
    </div>
    <div
      class="helpText"
    >
      Support custom variables: "current() + 2d 2h"
      <span
        class="bp3-popover-wrapper helpIcon"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="question"
            tabindex="0"
          >
            <svg
              height="14"
              viewBox="0 0 14 14"
              width="14"
            >
              <path
                d="M7 .1C3.2.1.1 3.2.1 7s3.1 6.9 6.9 6.9 6.9-3.1 6.9-6.9S10.8.1 7 .1zm0 1.2c3.2 0 5.7 2.6 5.7 5.7s-2.6 5.7-5.7 5.7S1.3 10.2 1.3 7 3.8 1.3 7 1.3zm0 2.3c-1.3 0-2.3 1-2.3 2.3h1.1c0-.6.5-1.1 1.1-1.1S8 5.3 8 5.9c0 .4-.3.8-.7 1h-.1c-.4.2-.8.6-.8 1.1v.7h1.1V8l.2-.1c.9-.3 1.5-1.1 1.5-2.1.1-1.2-.9-2.2-2.2-2.2zm-.6 5.7v1.1h1.1V9.3H6.4z"
                fill="#9AA5B5"
              />
            </svg>
          </span>
        </span>
      </span>
       
    </div>
  </div>
</div>
`;

exports[`<DateInput/> tests renders with value  1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="MM/DD/YYYY"
          type="text"
          value="01/23/2020"
        />
      </div>
    </div>
    <span
      class="bp3-popover-wrapper calendarIcon"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-calendar StyledProps--main"
            icon="calendar"
          >
            <svg
              data-icon="calendar"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                calendar
              </desc>
              <path
                d="M11 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1s-1 .4-1 1v1c0 .5.4 1 1 1zm3-2h-1v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H6v1c0 1.1-.9 2-2 2s-2-.9-2-2V1H1c-.6 0-1 .5-1 1v12c0 .6.4 1 1 1h13c.6 0 1-.4 1-1V2c0-.6-.5-1-1-1zM5 13H2v-3h3v3zm0-4H2V6h3v3zm4 4H6v-3h3v3zm0-4H6V6h3v3zm4 4h-3v-3h3v3zm0-4h-3V6h3v3zM4 3c.6 0 1-.5 1-1V1c0-.6-.4-1-1-1S3 .4 3 1v1c0 .5.4 1 1 1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
    </span>
  </div>
</div>
`;

exports[`<DateInput/> tests renders with value and allow variables 1`] = `
<div>
  <div
    class="dateInput"
  >
    <div
      class="main success"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="MM/DD/YYYY"
          style="padding-right: 0px;"
          type="text"
          value="current() + 2h"
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-tick StyledProps--main"
            icon="tick"
          >
            <svg
              data-icon="tick"
              height="12"
              viewBox="0 0 16 16"
              width="12"
            >
              <desc>
                tick
              </desc>
              <path
                d="M14 3c-.28 0-.53.11-.71.29L6 10.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l8-8A1.003 1.003 0 0014 3z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="helpText"
    >
      Support custom variables: "current() + 2d 2h"
      <span
        class="bp3-popover-wrapper helpIcon"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="question"
            tabindex="0"
          >
            <svg
              height="14"
              viewBox="0 0 14 14"
              width="14"
            >
              <path
                d="M7 .1C3.2.1.1 3.2.1 7s3.1 6.9 6.9 6.9 6.9-3.1 6.9-6.9S10.8.1 7 .1zm0 1.2c3.2 0 5.7 2.6 5.7 5.7s-2.6 5.7-5.7 5.7S1.3 10.2 1.3 7 3.8 1.3 7 1.3zm0 2.3c-1.3 0-2.3 1-2.3 2.3h1.1c0-.6.5-1.1 1.1-1.1S8 5.3 8 5.9c0 .4-.3.8-.7 1h-.1c-.4.2-.8.6-.8 1.1v.7h1.1V8l.2-.1c.9-.3 1.5-1.1 1.5-2.1.1-1.2-.9-2.2-2.2-2.2zm-.6 5.7v1.1h1.1V9.3H6.4z"
                fill="#9AA5B5"
              />
            </svg>
          </span>
        </span>
      </span>
       
    </div>
  </div>
</div>
`;
